src/gmx/context.py

```diff
@@ -157,91 +157,92 @@ def __init__(self, work):
-            assert isinstance(params, dict)
-            kwargs = {name: params[name] for name in params}
-            self.args = kwargs['args']
-            self.kwargs = kwargs['kwargs']
-
-            # The builder can hold an updater to be provided to the subscriber at launch. The updater is
-            # a function reference that the user provides to perform a desired periodic action. Not sure
-            # how this will work in the future. Allowing arbitrary Python code to be provided during job
-            # configuration would be the hardest thing. The updater could be another user-provided operation.
-            # Hopefully we can provided some canned behaviors that can be selected with element parameters.
-            # Maybe both.
-            self.updater = None
-
-        def add_subscriber(self, builder):
-            # At this point, we could find out how this data will be used (e.g. subscribing ranks)
-            if self.subscriber is None:
-                self.subscriber = builder
-            else:
-                raise exceptions.ApiError("This element does not support multiple consumers")
-
-        def build(self, dag):
-            # Either the builder needs to get the key for the subscribed node(s) so that we can
-            # create the edge now, or this builder needs to provide subscribers with the key of this
-            # DAG node so that the subscriber can create the edge. I think I prefer the latter, so that
-            # edge creation is tied to the binding process of a new node with its upstream nodes.
-            nodename = self.name
-            dag.add_node(nodename)
-            self.node = dag.nodes[nodename]
-
-            self.node['launch'] = self.__launch
-            # To avoid ambiguity, let's assert that only nodes that are active at launch will have data.
-            self.node['data'] = None
-            self.node['comm'] = self.comm
-            self.subscriber.input_nodes.append(nodename)
-            self.node['launch'] = self.__launch
-
-        def __launch(self, rank=None):
-            """Create the shared data resource for subscribed builders.
-
-            This is the place to provide subscribers with API references for interacting with
-            the shared data facility.
-            """
-            import numpy
-            data = numpy.empty(*self.args, **self.kwargs)
-            self.node['data'] = data
-            self.subscriber.shared_data_updater = self.updater
-            self.subscribinging_ranks = list(range(self.subscriber.width))
-
-            # Later, we can offload consumer responsibilities and the updater function to this runner,
-            # but we aren't there yet.
-            runner = None
-            return runner
-
-    builder = Builder(element)
-    return builder
+# def shared_data_maker(element):
+#     """Make a shared data element for use by dependent nodes.
+#
+#     Note: outdated. This code is in limbo and should not be used.
```

Comment:
What is the meaning of "in limbo" here?  Deprecated?  In the process of being refactored?

Reply:
Ah. Thanks. I should have tracked that better. There is a lot of code in the various repos that is currently unused or completely dead. I'm in the process of removing the dead code where it is untested, but for unused code that is tested, I have been content to make minimal changes. For this particular section, I think the code was tested and error free, but unused and inconsistent with forward-looking design. I'll take a closer look at it before letting it propagate further.

To do

- [ ] make better use of "to do"s
- [ ] remove shared_data_maker and test code, discussion, to one or more of #34, #36, #37, #45, #69

Reply:
There's not really a good place to stash ideas for posterity that I can find with GitHub. Unless maybe that's what Gist is for? I would want to stash the commented out bits of context.py and of test_mpiarraycontext.py. As it is, I tried to be more consistent with the commented out code and to include comments referring to tracked issues with commit 848e5c1. If you have other ideas, please let me know.